tools/publisher/src/cargo.rs

```diff
@@ -5,82 +5,72 @@
-pub fn publish_task(crate_path: &Path) -> Cmd {
-    cmd!["cargo", "publish", "--jobs", "1"].working_dir(crate_path)
+    /// Returns a plan string that can be output to the user to describe the command.
+    fn plan(&self) -> Option<Cow<'static, str>>;
 }
 
-#[derive(Default)]
-pub struct Cmd {
-    parts: Vec<String>,
-    working_dir: Option<PathBuf>,
+/// Confirms that cargo exists on the path.
+pub fn confirm_installed_on_path() -> Result<()> {
+    handle_failure(
+        "discover cargo version",
+        &Command::new("cargo")
+            .arg("version")
+            .output()
+            .context("cargo is not installed on the PATH")?,
+    )
+    .context("cargo is not installed on the PATH")
 }
 
-impl Cmd {
-    fn new() -> Cmd {
-        Default::default()
-    }
-
-    fn push(&mut self, part: impl Into<String>) {
-        self.parts.push(part.into());
-    }
-
-    fn working_dir(mut self, working_dir: impl AsRef<Path>) -> Self {
-        self.working_dir = Some(working_dir.as_ref().into());
-        self
-    }
-
-    /// Returns a plan string that can be output to the user to describe the command.
-    pub fn plan(&self) -> String {
-        let mut plan = String::new();
-        if let Some(working_dir) = &self.working_dir {
-            plan.push_str(&format!("[in {:?}]: ", working_dir));
-        }
-        plan.push_str(&self.parts.join(" "));
-        plan
-    }
+/// Returns (stdout, stderr)
+fn output_text(output: &Output) -> (String, String) {
+    (
+        String::from_utf8_lossy(&output.stdout).to_string(),
+        String::from_utf8_lossy(&output.stderr).to_string(),
+    )
+}
 
-    /// Runs the command asynchronously.
-    pub async fn spawn(mut self) -> Result<Output> {
-        let working_dir = self
-            .working_dir
-            .take()
-            .unwrap_or_else(|| std::env::current_dir().unwrap());
-        let mut command: Command = self.into();
-        tokio::task::spawn_blocking(move || Ok(command.current_dir(working_dir).output()?)).await?
+fn handle_failure(operation_name: &str, output: &Output) -> Result<(), anyhow::Error> {
+    if !output.status.success() {
+        return Err(capture_error(operation_name, output));
     }
+    Ok(())
 }
 
-impl From<Cmd> for Command {
-    fn from(cmd: Cmd) -> Self {
-        assert!(!cmd.parts.is_empty());
-        let mut command = Command::new(&cmd.parts[0]);
-        for i in 1..cmd.parts.len() {
-            command.arg(&cmd.parts[i]);
-        }
-        command
-    }
+fn capture_error(operation_name: &str, output: &Output) -> anyhow::Error {
+    let message = format!(
+        "Failed to {}:\nStatus: {}\nStdout: {}\nStderr: {}\n",
+        operation_name,
+        if let Some(code) = output.status.code() {
+            format!("{}", code)
+        } else {
+            "Killed by signal".to_string()
+        },
+        String::from_utf8_lossy(&output.stdout),
+        String::from_utf8_lossy(&output.stderr)
```

Comment:
when there are a lot of params, named params are a little easier to read IMO:
```suggestion
        "Failed to {name}:\nStatus: {status}\nStdout: {stdout}\nStderr: {stderr}\n",
        name = operation_name,
        status = if let Some(code) = output.status.code() {
            format!("{}", code)
        } else {
            "Killed by signal".to_string()
        },
        stdout = String::from_utf8_lossy(&output.stdout),
        stderr = String::from_utf8_lossy(&output.stderr)
```